From 1e51de5b91ff03413be83a8921b00972e2011b7f Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:25:01 -0400
Subject: [tools] perf/trace: Beautify epoll_ctl 'op' arg

Message-id: <1408462094-14194-133-git-send-email-jolsa@redhat.com>
Patchwork-id: 88125
O-Subject: [PATCH RHEL7.1 BZ1131394 132/325] perf trace: Beautify epoll_ctl 'op' arg
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit eac032c54be48e899a7be2b8db4b746be1b51748
Author: Arnaldo Carvalho de Melo <acme@redhat.com>
Date: Fri Sep 20 11:27:32 2013 -0300

description
===========
[root@sandy ~]# perf trace -e epoll_ctl
 2.490 (0.003 ms): systemd-logind/586 epoll_ctl(epfd: 10, op: ADD, fd: 24, event: 0x7fff22314ef0) = 0
 2.621 (0.003 ms): systemd-logind/586 epoll_ctl(epfd: 10, op: DEL, fd: 24                       ) = 0
 2.833 (0.010 ms): systemd-logind/586 epoll_ctl(epfd: 10, op: ADD, fd: 24, event: 0x7fff22314cd0) = 0
 2.953 (0.002 ms): systemd-logind/586 epoll_ctl(epfd: 10, op: DEL, fd: 24                       ) = 0
 3.118 (0.002 ms): systemd-logind/586 epoll_ctl(epfd: 10, op: ADD, fd: 24, event: 0x7fff22314d20) = 0
 4.762 (0.002 ms): systemd-logind/586 epoll_ctl(epfd: 10, op: DEL, fd: 24                       ) = 0
^C[root@sandy ~]#
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/builtin-trace.c b/tools/perf/builtin-trace.c
index a8c99a4..c712a78 100644
--- a/tools/perf/builtin-trace.c
+++ b/tools/perf/builtin-trace.c
@@ -232,6 +232,9 @@ static size_t syscall_arg__scnprintf_futex_op(char *bf, size_t size, struct sysc
 
 #define SCA_FUTEX_OP  syscall_arg__scnprintf_futex_op
 
+static const char *epoll_ctl_ops[] = { [1] = "ADD", "DEL", "MOD", };
+static DEFINE_STRARRAY(epoll_ctl_ops);
+
 static const char *itimers[] = { "REAL", "VIRTUAL", "PROF", };
 static DEFINE_STRARRAY(itimers);
 
@@ -543,6 +546,9 @@ static struct syscall_fmt {
  { .name     = "brk",     .hexret = true,
    .arg_scnprintf = { [0] = SCA_HEX, /* brk */ }, },
  { .name     = "connect",    .errmsg = true, },
+ { .name     = "epoll_ctl",  .errmsg = true,
+   .arg_scnprintf = { [1] = SCA_STRARRAY, /* op */ },
+   .arg_parm  = { [1] = &strarray__epoll_ctl_ops, /* op */ }, },
  { .name     = "eventfd2",   .errmsg = true,
    .arg_scnprintf = { [1] = SCA_EFD_FLAGS, /* flags */ }, },
  { .name     = "fcntl",     .errmsg = true,
-- 
1.7.1